common/src/main/java/org/conscrypt/ConscryptEngineSocket.java

```diff
@@ -447,14 +449,22 @@ public final void close() throws IOException {
         }
 
         try {
-            // Close the underlying socket.
-            super.close();
-        } finally {
             // Close the engine.
             engine.closeInbound();
             engine.closeOutbound();
-            
-            // Release any resources we're holding
+            // Closing the outbound direction of a connected engine will trigger a TLS close
+            // notify, which we should try and send.
+            // If we don't, then closeOutbound won't be able to free resources because there are
+            // bytes queued for transmission so drain the queue those and call closeOutbound a
+            // second time.
+            if (previousState >= STATE_HANDSHAKE_STARTED && out != null) {
```

Comment:
seems to me that out != null check belongs to drainOutgoingQueue()

Reply:
I thought so too, but it can be null when if the handshake throws before it starts, e.g. due to invalid TLS params, which fortunately we have a test for. It should never be null after the handshake is started though so I removed the check here to prevent it masking bugs.